block-producer/src/batch_builder/mod.rs

```diff
@@ -18,66 +24,103 @@ mod tests;
+            }
+
+            SimpleSmt::with_leaves(
+                CREATED_NOTES_SMT_DEPTH,
+                created_notes.into_iter().enumerate().map(|(idx, note_hash)| {
+                    (
+                        idx.try_into().expect("already checked not for too many notes"),
+                        note_hash.into(),
+                    )
+                }),
+            )?
+        };
+
+        Ok(Self {
+            updated_accounts,
+            produced_nullifiers,
+            created_notes_smt,
+        })
     }
 
     /// Returns an iterator over account ids that were modified in the transaction batch, and their
     /// corresponding initial hash
     pub fn account_initial_states(&self) -> impl Iterator<Item = (AccountId, Digest)> + '_ {
-        self.txs.iter().map(|tx| (tx.account_id(), tx.initial_account_hash()))
+        self.updated_accounts
+            .iter()
+            .map(|(account_id, account_states)| (*account_id, account_states.initial_state))
     }
 
     /// Returns an iterator over account ids that were modified in the transaction batch, and their
     /// corresponding new hash
     pub fn updated_accounts(&self) -> impl Iterator<Item = (AccountId, Digest)> + '_ {
-        self.txs.iter().map(|tx| (tx.account_id(), tx.final_account_hash()))
-    }
-
-    /// Returns the script root of all consumed notes
-    pub fn consumed_notes_script_roots(&self) -> impl Iterator<Item = Digest> + '_ {
-        let mut script_roots: Vec<Digest> = self
-            .txs
+        self.updated_accounts
             .iter()
-            .flat_map(|tx| tx.consumed_notes())
-            .map(|consumed_note| consumed_note.script_root())
-            .collect();
-
-        script_roots.sort();
-
-        // Removes duplicates in consecutive items
-        script_roots.into_iter().dedup()
+            .map(|(account_id, account_states)| (*account_id, account_states.final_state))
     }
 
     /// Returns the nullifier of all consumed notes
     pub fn produced_nullifiers(&self) -> impl Iterator<Item = Digest> + '_ {
-        self.txs
-            .iter()
-            .flat_map(|tx| tx.consumed_notes())
-            .map(|consumed_note| consumed_note.nullifier())
+        self.produced_nullifiers.iter().cloned()
     }
 
     /// Returns the hash of created notes
     pub fn created_notes(&self) -> impl Iterator<Item = Digest> + '_ {
-        self.txs
-            .iter()
-            .flat_map(|tx| tx.created_notes())
-            .map(|note_envelope| note_envelope.note_hash())
+        self.created_notes_smt.leaves().map(|(_key, leaf)| leaf.into())
+    }
+
+    /// Returns the root of the created notes SMT
+    pub fn created_notes_root(&self) -> Digest {
+        self.created_notes_smt.root()
     }
 }
 
+/// Stores the initial state (before the transaction) and final state (after the transaction) of an
+/// account
+struct AccountStates {
```

Comment:
minor nit: `AccountStates` name feels a bit too general - but I don't really have a better suggestion (one option that came to mind was `AccountTransition` but not sure if it is better).